### PR TITLE
[BEAM-2576] Split Beam portability framework into construction, job, and execution modules

### PR DESCRIPTION
**Pull request by kennknowles**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

The Beam portability framework has a lot of simple pieces that can be modularized more usefully than "Runner API" and "Fn API".

Here are some semantic categories that should remain stable, and make sense to depend on separately:

 - Modeling of a Pipeline and pieces of it
 - Standardized transform payloads
   - primitives
   - non-primitives
   - execution-time only transforms
 - Standardized UDFs
 - Job submission and management (Job API)
 - Artifact upload and retrieval (Artifact API) (could separate upload and retrieval)
 - Model of bundle processing instructions (ProcessBundleDescriptor)
 - Standardized execution-only instruction payloads (like data plane read instructions)
 - Setting up the SDK harness (Provision API)
 - Control plane for bundle processing (FnControl)
 - Multi-purpose data streaming (FnData)
 - Logging
 - ... more later, maybe ...

This PR is a first move in this direction, organizing the components according to when in a pipeline's lifecycle they are used.

 - **Construction**: building a pipeline and specifying transforms. Basically just data structures, not APIs. Code doing this work should only depend on this module.
 - **Job**: submission and management of a pipeline, including artifacts, etc. Construction code should not depend on this module.
 - **Execution**: SDK harness management and UDF invocation. Only runner internals and SDK harnesses should depend on this module.

What this PR does is move the existing protos into the right maven modules, but does not split them yet.

### CODE REVIEW

**Comment by kennknowles** [comment]
R: @robertwb 

(anyone's comments welcome, like @lukecwik @tgroh @aljoscha, @herohde, @tweise. I'm just limiting reviewer resources consumed)

**Comment by aljoscha** [comment]
I think splitting it up this way makes a lot of sense. I recently gave a talk at Flink Forward Berlin about the compatibility efforts and I roughly explained the API with those separations but I didn't take the extra further step of thinking how it should be organised in the actual code.

File: model/job/src/main/proto/beam_artifact_api.proto [hunk 1]
```diff
@@ -23,9 +23,9 @@
 
 syntax = "proto3";
 
-package org.apache.beam.runner_api.v1;
+package org.apache.beam.model.job.v1;
 
-option java_package = "org.apache.beam.sdk.common.runner.v1";
+option java_package = "org.apache.beam.model.job.v1";
```
Comment (herohde) [inline model/job/src/main/proto/beam_artifact_api.proto]: Artifact should strictly speaking be split in 3 now: data structures, stage (submission) and retrieval (execution).

**Comment by herohde** [comment]
This split makes a lot of sense and will reduce confusion down the line. Not to be that guy, but the names "job" and "construction" seem slightly off to me -- I'm thinking more of them as just "model" and "submission"/"management", say.

**Comment by kennknowles** [comment]
I think now is the perfect time to have a naming discussion :-)

**Construction -> ??**

I think "model" is a bit general as a replacement for "construction". It _is_ pretty much what we know as the "Beam programming model". I want to explicitly make clear that there is separately an execution model and a job submission/management model.

I toyed with "pipeline" but opted for something more broad in order to also lump the primitive payloads and standardized payloads into the same artifact. There's no technical reason they need to be in one artifact, but just keeping the artifact count lower because maven is heavy on boilerplate.

Could definitely split into `beam-model-pipeline`, `beam-model-primitives`, `beam-model-standard-transforms`, and `beam-model-standard-fns`.

**Job -> ??**

No strong opinion, but isn't "job" the thing that you are submitting and managing?

**Comment by tgroh** [comment]
`standard-transforms` would contain, for example, `Combine`?

I'm pretty happy with `job` or `job-management`; submission is to specific (because you can also get out information about running jobs, update, drain, cancel, etc.

I don't know how useful I find `model` as a common prefix, especially when the subpackages are things like 'construction' and 'job'. Something like `org.apache.beam.model.pipeline...` mostly makes sense to me, and I suppose to some extent `...model.execution...` does as well. I don't think `...model.job...` scans particularly well, however. It may be because the job APIs are the only one where there are two top-level systems communicating (at least in how I comprehend the three divisions here)

**Comment by herohde** [comment]
For "job", I was more expecting something contrasting "execution". I agree that jobs are what we submit and manage, but also what we execute.

For "construction". the split names seem clearer. I don't have a better suggestion and don't feel too strongly about it, but "construction" suggests something active as opposed to passive data structures to me.

**Comment by kennknowles** [comment]
Maybe "execution" is the worst of all. Maybe `fn` is still the best for this. Beyond what is currently in `beam_fn_api` it would also have provision and other bits pertinent to setting up a harness and managing it. Certainly its whole point is to invoke UDFs.

**Comment by kennknowles** [comment]
@tgroh yes, `standard-transforms` would contain `Combine` and perhaps some more like `Create` and `WriteFiles` (most IOs should manage their own payloads, but this one is in the core SDK). The distinction between primitives and standardized transforms is perhaps not critical here.

**Comment by lukecwik** [comment]
Based upon the list of parts, I grouped them into the three areas I believe they should go:
- beam-model-core:
  -  Modeling of a Pipeline and pieces of it
  - Standardized transform payloads
  - primitives
  - non-primitives (only ones which are shipped with Apache Beam, users creating their own should package and ship their own)
  - Standardized UDFs
- beam-model-job-management:
  - Job submission and management (Job API)
  - Artifact upload
- beam-model-job-execution:
  - Artifact retrieveal
  - Model of bundle processing instructions (ProcessBundleDescriptor)
  - Standardized execution-only instruction payloads (like data plane read instructions)
  - Setting up the SDK harness (Provision API)
  - Control plane for bundle processing (FnControl)
  - Multi-purpose data streaming (FnData)
  - Logging
  - execution-time only transform payloads

**Comment by kennknowles** [comment]
@lukecwik awesome - that's exactly what I've got, except the artifact upload vs retrieval split. For this PR I just want to move proto files without splitting them, but afterwards I will do a little bit towards the breakdown you outlined.

**Comment by robertwb** [comment]
That division makes a lot of sense to me, except having Artifact Upload and Artifact Retrieval separate seems a bit odd (but I could live with that, as long as they share a common definition of Artifact).

**Comment by lukecwik** [comment]
Yes, I was thinking the same thing now that you mentioned it that the a common Artifact definition should fall under beam-model-core.

**Comment by tgroh** [comment]
It's a bit of a strange place for it, given that it has no use until the Job API, which does not appear to be part of the core model

**Comment by lukecwik** [comment]
Artifact upload is part of job management though.

**Comment by robertwb** [comment]
We should think hard about whether the notion of artifacts (which seem
similar in spirit to environments) should be in the core model, especially
if they're per-step or anything like that. I agree that submitting
artifacts belongs as part of the job API (though perhaps not distinct from
submitting the job itself) and fetching them part of the execution API.

**Comment by kennknowles** [comment]
Agree with @robertwb that there's a decent chance the TBD `Environment` bit on UDFs may reference things like artifacts. It seems like a bit much to include the whole classpath, or to narrow the list down effectively, but there may well be special cases, like when the user specifies additional resources to stage. I think "core" might be a loaded term that makes us more hesitant to include things than we need to be.

**Comment by kennknowles** [comment]
Since this is a pretty conflict-ridden PR due to imports, can we go with an LGTM on the basic modules so I can asynchronously fix conflicts and merge?

Specifically, based on this feedback, how about:

 - `model/pipeline`
 - `model/job-management`
 - `model/fn`

These are just package renames from what is already here.

**Comment by robertwb** [comment]
+1 on getting basic modules in to avoid continuous rebasing.

One though: what about

    model/worker-management

instead of fn?

**Comment by kennknowles** [comment]
So far we haven't really had a notion of a worker in Beam - every runner has their own sort of worker, sometimes doing runner-specific stuff. I'd hate to cause confusion there.

I guess I like `fn` because it has to do with invoking DoFns, WindowFns, etc. But it pretty much entails most aspects of execution since you need to fuse and code/encode, etc. Hence my first draft being `model/execution`, but that wasn't specific about exactly what was being executed.

**Comment by kennknowles** [comment]
`fn-execution`? maybe that is just extra characters. On the other hand, folks like `job-management` more than `job` (I still like the pithier `job`, personally)

**Comment by robertwb** [comment]
+1 to fn-execution. (We do have the concept of a "SDK worker," but I agree
the term worker is way overloaded. I'm all for pithy, but I think in this
context a little verbosity is worth the clarity.

**Comment by kennknowles** [comment]
So, I'll go ahead with `model/pipeline`, `model/job-management` and `model/fn-execution`.

**Comment by kennknowles** [comment]
Ah, the issue now is that the module `beam-runners-core-construction` depends on `beam-model-job-management` but it is not in the reactor. So it may be that to get maven to function correctly we'll need to hand-curate the list of modules.

**Comment by kennknowles** [comment]
retest this please

**Pull request merged**